mysys/lf_hash.c

```diff
@@ -111,13 +111,13 @@ static int l_find(LF_SLIST **head, CHARSET_INFO *cs, uint32 hashnr,
 
     cur_hashnr= cursor->curr->hashnr;
     cur_keylen= cursor->curr->keylen;
-    cur_key= cursor->curr->key;
+    cur_key= my_atomic_loadptr((void *) &cursor->curr->key);
 
     do {
       link= cursor->curr->link;
       cursor->next= PTR(link);
       lf_pin(pins, 0, cursor->next);
-    } while (link != cursor->curr->link && LF_BACKOFF);
+    } while (link != (intptr) my_atomic_loadptr((void *) &cursor->curr->link) && LF_BACKOFF);
```

Comment:
Feels like it should be `MY_MEMORY_ORDER_RELAXED`, and another load of `link` a few lines above - `MY_MEMORY_ORDER_ACQUIRE`.

And corresponding code that publishes node via `link` should issue `MY_MEMORY_ORDER_RELEASE`.

Reply:
Making the load of `LF_SLIST::link` atomic ACQUIRE instead of SEQ_CST is also fine. (Both translate into the same ARM binary.)
In the case at hand (reordering the load of key to happen after the loop), it does not matter which of the two loads of `LF_SLIST::link` is done atomically with ACQUIRE ordering.

However, in our microbenchmark of lf-hash we experience roughly a 20% performance degradation if the first (loop-body) load is made atomic ACQUIRE instead making the later (exit-condition) load atomic ACQUIRE. We do not know if this translates into a similar performance difference for the lf-hash in the main code base, though.

Reply:
Yes, there seem to be no difference which load of the `link` is going to issue `ACQUIRE`. It just has to be issued before accessing members of that node.

The reason for 20% decline is not evident to me. But if that's the case it'd make sense to issue acquire along with the second load. Although if I were owner of this issue, I'd put some extra effort in figuring this out. The purpose of the first load is "load", while the purpose of the second load is "make sure it hasn't changed in the meantime". Issuing `ACQUIRE` along with the first load is easier to interpret.

Reply:
Why are these barriers needed? Two loads are separated with lf_pin() which is store with a SEQ_CST semantics. A CPU shouldn't be reordering loads across it, should it?

Reply:
The performance difference might be an artifact or depend on the actual workload. Running lf-t a couple of times did not result in a significant difference. From the theoretical point of view it does not seem to be obvious to argue for one or the other solution, performance wise.

Loop condition acquire:
```bash
$ time for i in `seq 1000`; do ./lf-t.acq-cond.bin; done > /dev/null
real	8m11.859s
user	208m1.040s
sys	0m3.343s
```

Loop body acquire:
```bash
$ time for i in `seq 1000`; do ./lf-t.acq-loop.bin; done > /dev/null
real	8m15.403s
user	209m4.627s
sys	0m3.663s
```

Reply:
@vuvova new node is published via `LF_HASH` by updating `LF_SLIST::link` of preceding element. That's the only way non-dummy node can be accessed.

In a publisher thread we need to make sure memory of the node was "committed" before it got published via `LF_SLIST::link`, thus `RELEASE`. Corresponding `ACQUIRE` has to be issued by the loader thread, which is going to access this node by traversing `LF_SLIST::link` of the preceding element.

It follows common release-acquire model. Makes sense?

Barrier issued along with `lf_pin()` is irrelevant here, because it is attached to a different variable.

Reply:
@mbeck21 nice, thanks for an update!

Reply:
@vuvova An atomic store SC is not the same as a relaxed store with a full memory barrier. In fact on ARM a SEQ_CST store as well as an ACQUIRE store are translated into 'STLR'.

From the ARM architecture /memory model / [barrier documentation](https://developer.arm.com/documentation/100941/0100/Barriers) it states that:

> All loads and stores preceding an STLR that match the shareability domain of the target address must be observed before the STLR.

So loads and stores appearing after STLR in program order are allowed to be executed before STLR. 

l_find looks like this in program order:

1. RLX R key
2. RLX R link
3. REL W pin
4. RLX R link

It thus is possible for the relaxed read of link (at the exit condition) (4) to happen before (3) store release.

1. RLX R key
2. RLX R link
3. RLX R link
4. REL W pin

Further, as relaxed accesses can be reorder if there is no dependency preventing it, it may happen that:

1. RLX R link
2. RLX R link
3. RLX R key
4. REL W pin

Would it help you to check with the official [herd7 arm verification tool](http://diy.inria.fr/www/?record=aarch64)?

Reply:
@vuvova, a few more thoughts on this loop. If we get it simplified to something like:
```C++
1. key= cursor->curr->key
loop:
2. link= cursor->curr->link
3. pins[0]= link
4. check= cursor->curr->link
   if (link != check)
     goto loop;
```
The order constraints are:
* `1.` must not get reordered past `4.` (or we get this bug)
* `2.` and `3.` are naturally ordered (because the latter depends on the former)
* `4.` must not get reordered before `3.` (if it is, node can get removed before it is pinned)
*  in other words `2.`, `3.` and `4.` must be executed in the program order.

Additional order constraints:
* all concurrent stores to node memory accessible via `link` (next node) must be completed by `4.`, thus `4.` or `1.` has to issue `ACQUIRE` or `CONSUME` or `SEQ_CST`
* store to `pins[0]` must be completed before previous store to `pins[1]` (safe pin transition of current node), thus it has to issue `RELEASE` or `SEQ_CST`.

To sum up all the above, here is an option that covers all of the above constraints and has least overhead:
```C++
cur_key= my_atomic_loadptr_explicit((void *) &cursor->curr->key, MY_MEMORY_ORDER_RELAXED);
do {
  link= (intptr) my_atomic_loadptr_explicit((void *) &cursor->curr->link,
                                            MY_MEMORY_ORDER_RELAXED);
  cursor->next= PTR(link);
  lf_pin(pins, 0, cursor->next);
} while (link != (intptr) my_atomic_loadptr((void *) &cursor->curr->link) && LF_BACKOFF);
```

Alas, given the number of concurrent variables involved, `LF_HASH` has to operate using `SEQ_CST` model. :(

Thanks @mbeck21 for your reordering matrices and thanks @vuvova for planting a seed of doubt. I must admit I missed that famous store-load problem (reordering `4.` before `3.`), that can only be solved with `SEQ_CST`.

Reply:
Indeed this is a nice optimization that should work on Arm. However, I am not sure that it will work on Power and other WMM architectures. At least it is not ok according to the C11 memory model.

Here is an example to demonstrate that with [CppMem](http://svr-pes20-cppmem.cl.cam.ac.uk/cppmem).

```c
int main() {
    atomic_int key = 0;
    atomic_int link = 0;
    atomic_int pin = 0; 
    int old = 0;
    {{{ { /* thread 1 */
          cas_strong_explicit(link, old, 1, memory_order_seq_cst, memory_order_seq_cst);
          key.store(1, memory_order_seq_cst);
        }
    ||| { /* thread 2 */
          r1 = key.load(memory_order_relaxed).readsvalue(1);
          pin.store(1, memory_order_release);
          r2 = link.load(memory_order_seq_cst).readsvalue(0);
        }
    }}};
    return 0; }
```

If you run this, you'll find that the C11 memory model allows a *consistent* execution of both threads such that the `key` update (`readsvalue(1)`) is seen but the update of `link` is not (`readsvalue(0)`). The reason is that there is no read of pin in thread 1 (between CAS and store) which would create a synchronize-with edge. Even making the store to `pin` *memory_order_seq_cst* is not sufficient (since it kinds of map to *memory_order_release*). The only solution is to make the load of `key` to be *memory_order_acquire* or to insert an `atomic_thread_fence(memory_order_acquire)` or stronger between the load of the `key` and the load of the `link`.

Note that the CAS and store in thread 1 issue *memory_order_seq_cst* so that we separate the concerns and avoid any issues on thread 1.

Reply:
> The reason is that there is no read of pin in thread 1 (between CAS and store) which would create a synchronize-with edge. 

Yeah, that's what I was unsure about. If that's the case, `key` has to follow full release-acquire model. :(
That is `ACQUIRE` here and `RELEASE` in `add_to_purgatory()`.